umbraco/Umbraco-CMS
# Use SnippetCollection to when working with snippets

## Details
Since the partial view macro snippets have been moved to Umbraco.Core as embedded resources, we can use a collection builder to get a collection of all embedded and custom snippets (added to `~\Umbraco.Web.UI\umbraco\PartialViewMacros\Templates` folder by a package, for example) - `PartialViewMacroSnippetCollection`.

There is a collection for the Partial View Snippets as well (`PartialViewSnippetCollection`), which contains just the embedded snippets (_with a few exceptions*_).
&nbsp;&nbsp;&nbsp;&nbsp; _\* "Gallery", "List Child Pages From Changeable Source", "List Child Pages Ordered By Property", "List Images From Media Folder"_


## Test
- Try creating a new Partial View Macro File from snippet and observe that you get the same amount of elements as before;
  - Create a new PVM file from a snippet and compare its contents with a snippet from an Umbraco solution before this PR;
- Try creating a new Partial View File from snippet and observe that you get the same amount of elements as before;
  - Same snippets as for a Partial View **Macro** but without "**Gallery**", "**List Child Pages From Changeable Source**", "**List Child Pages Ordered By Property**", "**List Images From Media Folder**";
  - Create a new PV file from a snippet and compare its contents with a snippet from an Umbraco solution before this PR - that should be the same;
- Create a new Razor view and add it to  `~\Umbraco.Web.UI\umbraco\PartialViewMacros\Templates` folder (_code snippet below_);
  - Observe that it has been added to the end of the available snippets (just for the partial view macro snippets);
-  Create another Razor view with the same name as any of the embedded snippets - ex: Empty.cshtml (_code snippet below_);
   - Observe that the contents of an Empty snippet are the new ones (and not those from the embedded resource) - for a PVM. While the Empty snippet for a PV should contain the original Empty snippet contents;

❗**Note**: The snippets added to the `Templates` folder will be always listed after the embedded ones (when trying to create a PVM from snippet; they won't appear if you try to create a PV from snippet). The only exception is the _Empty.cshtml_. Eeven though the contents of it might be overwritten by a custom Empty.cshtml template, it will always be displayed on top of the list.

## Appendix

Index.cshtml: 
```c#
@inherits Umbraco.Cms.Web.Common.Macros.PartialViewMacroPage

@{
}
```

Empty.cshtml: 
```c#
@inherits Umbraco.Cms.Web.Common.Macros.PartialViewMacroPage

@{
    // NEW CONTENT
}
```
